File: pcseg/model/segmentor/fusion/rpvnet/range_lib/range_utils/src/denselize.cpp
#include <torch/torch.h>
#include <vector>
#include "denselize_gpu.h"

at::Tensor denselize_forward(
    const at::Tensor feat,
    const at::Tensor count_map,
    const at::Tensor pxpy
)
{
    int B = count_map.size(0);
    int W = count_map.size(2);
    int H = count_map.size(1);
    int C = feat.size(1);
    int N = pxpy.size(0);
    at::Tensor feature_map = torch::zeros({B,C,H,W}, at::device(pxpy.device()).dtype(at::ScalarType::Float));
    denselize_wrapper(N,C,H,W,feat.data_ptr<float>(),count_map.data_ptr<int>(),pxpy.data_ptr<int>(),feature_map.data_ptr<float>());
    return feature_map;    
}

at::Tensor denselize_backward(
    const at::Tensor top_grad,
    const at::Tensor count_map,
    const at::Tensor pxpy
)
{
    int B = count_map.size(0);
    int W = count_map.size(2);
    int H = count_map.size(1);
    int C = top_grad.size(1);
    int N = pxpy.size(0);
    at::Tensor bottom_grad = torch::zeros({N, C}, at::device(pxpy.device()).dtype(at::ScalarType::Float));
    denselize_grad_wrapper(N,C,H,W,top_grad.data_ptr<float>(),count_map.data_ptr<int>(),pxpy.data_ptr<int>(),bottom_grad.data_ptr<float>());
    return bottom_grad;
}

File: pcseg/model/segmentor/fusion/rpvnet/range_lib/range_utils/nn/functional/denselize.py
from torch.autograd import Function

import rangelib_cuda

__all__ = ['denselize']

class DenselizeGPU(Function):
    @staticmethod
    def forward(ctx, feat, count_map, pxpy):
        '''
        args:
            feat: torch.Float32 tensor N X C 
            count_map: int max batch size
            pxpy: torch.Int32 tensor N X 3 (mean batch_id, px, py)
                px range:0~w 
                py range:0~h
        return:
            outs:torch.Int32 tensor B C X H X W
        '''
        outs = rangelib_cuda.denselize_forward(feat, count_map, pxpy)
        ctx.for_backwards = (count_map.int().contiguous(), pxpy.int().contiguous())
        return outs

    @staticmethod
    def backward(ctx, top_grad):
        count_map, pxpy = ctx.for_backwards
        bottom_grad = rangelib_cuda.denselize_backward(top_grad.float().contiguous(),count_map,pxpy)
        return bottom_grad, None, None


denselize_gpu = DenselizeGPU.apply

def denselize(feat, count_map, pxpy):
    return denselize_gpu(feat, count_map, pxpy)
